### ISSUE TITLE
Ticket Caching Failing With Visual Studio 2017

### ISSUE THREAD

**Issue by hokiecoder** [apereo/dotnet-cas-client#54]
The TL;DR is that when running a CAS-enabled application in Visual Studio 2017 (IIS Express), the .NET cache provider (`System.Web.Caching.Cache`) is immediately expiring the new service tickets as soon as they are added to the cache. Thus, no service ticket is ever able to be validated. Below is how I arrived at this conclusion.

Recently my development computer at work bit the dust, prompting me to reformat and re-install Windows 10. I took that opportunity to also upgrade to Visual Studio 2017 Enterprise (from 2015 Enterprise). When I began working on one of my CAS-enabled applications in VS2017, I immediately got the problem of a redirect cycle between my application and the CAS server.

First thing I checked was that it wasn't another timing problem with the tolerance level for the SAML tickets, which it wasn't. No matter what value I set for ticketTimeTolerance, the redirects still occurred. I debugged about all I could think of in my application with no indication of why the SAML ticket validation was failing and thus redirecting back to the CAS server. Even turning on the tracing per the instructions in this project's README didn't shine light on the source of the problem.

I finally broke down and added the DotNetCasClient project to my solution so that I could step through the code as the process was happening. Eventually I noticed that the service ticket was being added to the cache, but in a subsequent call to retrieve the ticket from the cache, `null` was returned. I added some code for debugging to [InsertTicket()](https://github.com/apereo/dotnet-cas-client/blob/1db5d00f2eb10858b68d9076d381fcd67c4364f3/DotNetCasClient/State/CacheServiceTicketManager.cs#L95). I was able to see that, immediately after inserting the ticket into the cache the cache item count increased by one. However, after a few debugging steps, the cache count decreased by one. I read in the comments for [RemoveExpiredTickets()](https://github.com/apereo/dotnet-cas-client/blob/1db5d00f2eb10858b68d9076d381fcd67c4364f3/DotNetCasClient/State/CacheServiceTicketManager.cs#L62) that the cache will automatically remove any expired items, so this got me thinking that the problem was with the expiration date for the cache item being inserted. By pure luck I was able to notice that the dates for the forms authentication ticket were using UTC while the expiration date for the cache item was in local time. Furthermore, I noticed the comments for the `Insert()` method of `System.Web.Caching.Cache` recommend to always use UTC dates as issues can occur with daylight savings time. The particular problem I am having wasn't related to DST, but on a whim I changed the insert statement from:

`HttpContext.Current.Cache.Insert(GetTicketKey(casAuthenticationTicket.ServiceTicket), casAuthenticationTicket, null, expiration, Cache.NoSlidingExpiration);`

to:

`HttpContext.Current.Cache.Insert(GetTicketKey(casAuthenticationTicket.ServiceTicket), casAuthenticationTicket, null, expiration.ToUniversalTime(), Cache.NoSlidingExpiration);`

Immediately my application worked again. To double-check that the fix was correct I have created another dummy web application that consumes CAS in Visual Studio 2017. Using the local date/time, the redirect loop occurs. Modify it to use UTC date/time, the application works as expected.

I created the same dummy application in a copy of Visual Studio 2015 I have, targeting the same version of the .NET framework (v4.5.2). Using the same configuration settings for the client in web.config, I did the same tests, except this time the caching of the service tickets works regardless of using local or UTC time for expiration. As best as I can tell, the only difference between the projects is Visual Studio 2015 versus 2017. They target the same framework version, have all the same NuGet packages/versions, and use the same web.config settings.

Using those same dummy projects (one in 2015, the other in 2017) I basically copied the code from [ContainsTicket()](https://github.com/apereo/dotnet-cas-client/blob/1db5d00f2eb10858b68d9076d381fcd67c4364f3/DotNetCasClient/State/CacheServiceTicketManager.cs#L151) to list the contents of the cache on each page load. Using a special query string parameter, I could instruct the page load code to add a simple string value as an object to the cache with a +1 minute expiration date. In VS2017, using the local date for expiration, the string would be gone from the cache on the next load, even though the second request was only a few seconds after adding the object to the cache. Using the UTC date, the string would stay in the cache for a minute, then be removed. Switching to VS2015 yielded the expected result that it did not matter which form of the date/time was used for expiration. The string value would stay in cache for 1 minute, then expire.

According to the [MSDN Documentation for the Insert() method](https://msdn.microsoft.com/en-us/library/4y13wyk9(v=vs.110).aspx), the recommendation for the `absoluteExpiration` parameter value is for it to be a UTC date/time, not a local one. It is laughable, though, that the examples on the same page use `DateTime.Now`, against their own recommendation. In any case, I think it would be a good thing to change the expiration to a UTC date/time regardless of it being the absolute root cause of the problem I am experiencing.

I believe the change would need to be made at the point where the expiration date is set, which is when the forms authentication ticket is being created in [CreateFormsAuthenticationTicket()](https://github.com/apereo/dotnet-cas-client/blob/1db5d00f2eb10858b68d9076d381fcd67c4364f3/DotNetCasClient/CasAuthentication.cs#L1132-L1133) of `CasAuthentication.cs`. The `fromDate` and `toDate` variables can be set as UTC dates, and the client has no problem in VS2017. I will create a pull request soon with the suggested changes along with documentation as to why UTC dates are needed. In the meantime, if anyone wants to weigh in on why this is happening between VS2015 (and earlier) and VS2017, it would be appreciated!

**Comment by hokiecoder** [comment]
Another thing that I am curious about testing is whether or not the client, as is, experiences any issues on IIS on Windows Server 2016. I have only been using the client on up to Windows Server 2012 R2 without issue. But I wonder if whatever changed in VS2017 also changed in Windows Server 2016/IIS. When I can get my hands on IIS on 2012 I'll try it out. If anyone else is able to test this, I'd love to hear the results.

**Comment by danransom** [comment]
We're seeing the same thing with Windows 10 Creator update (1703) in all versions of Visual Studio. Creating a pull request with this fix ...

**Comment by webMan1** [comment]
Thank you! I've been looking for this solution for days. This really needs to be done. For now, I've found that a quick fix is changing the time zone on your computer to UTC. If you don't mind your clock being off by a couple of hours while you're testing, it works okay.

**Comment by phantomtypist** [comment]
@hokiecoder do you work at Virginia Tech?

**Comment by phantomtypist** [comment]
So here are my thoughts for this issue after discussing with my team.

- This bug, problem if you will, is introduced when .NET 4.7 is installed on the machine that this code (your web application) runs on.
- .NET 2.0 and 3.x applications are unaffected.  The only impact is on .NET 4.x applications.

The following thoughts are in preparation of planning for the future of this project...

- The nuget package will have two assemblies in it for this bug fix.
- When you install the nuget package, if your application targets the .NET 2.0/3.x framework, then you get the assembly compiled for that version of the framework and will contain the same code as today to maintain backwards compatibility.
- If you install the nuget package in an application that targets the .NET 4.x framework, you get this assembly compiled for that version of the framework.  Herein lies the reason:  I think it's easiest to swap out the cache provider with System.Runtime.Caching and not muck with the expiration date.
- This single nuget package, multiple targeted assemblies is a common thing in the .NET landscape.  This can be seen in many packages, including the most downloaded which is JSON.NET.
- This way there is no need to add/change web.config settings and things will just work out of the box without developer intervention.  Think of it as an "under the hood" change.

One of our future goals would be to introduce the concept of a plugable cache provider interface where you could use the default standard in-memory cache provider, the two described above, or a third party of your choosing.  One use case would be to allow for a distributed cache provider such as memcached or redis.  This will make it easier to move CAS enabled .NET applications to a load balanced environment, especially if "sticky sessions" are not enough.

**Comment by hokiecoder** [comment]
@phantomtypist Yes, I do work at Virginia Tech. Also, to your last comment, I am not sure we need to go through all the work of building for two different .NET targets. At least with my testing, fixing the ticket expiration date to be a UTC date does not cause any problems with earlier versions of .NET. It seems that the `System.Web.Caching.Cache` class previously could handle dates in either formats, but now has more strict requirements. I may be wrong about this, so testing the patch provided by **danransom** in pull request #55 against multiple .NET targets should be done.

As far as a pluggable model for the service ticket manager (aka dependency injection), the client is already set up for that using the `IServiceTicketManager` interface. The `CacheServiceTicketManager` implements this interface. It would be possible, and fairly easy, to write your own implementation for `System.Runtime.Caching`, using your own example. I am not very familiar with the Reflection technique that is used in the `Initialize()` method of `CasAuthentication` to create an instance of the service ticket manager you specify in the configuration, but it looks like it could load any class available at runtime to the assembly as the service ticket manager, provided it implements `IServiceTicketManager`.

**Comment by phantomtypist** [comment]
Just to clarify... did you test this fix against an application compiled in .NET 2/3.x?

**Comment by phantomtypist** [comment]
Ok, found out we probably don't need to change anything in the long term, maybe just temporarily... it's an actual bug in .NET 4.7's implementation of Cache.Insert.  I'll have to ask my contacts at MS if they can track down the status of this.

Ref: https://forums.asp.net/p/2123507/6145985.aspx

**Comment by phantomtypist** [comment]
But IMHO, we should switch out to the System.Runtime.Caching provider for the .NET 4.x targets (it's not in .NET 2/3.x because it was part of Enterprise Library at that point.)

I'm suggesting the dual compilation targets because if you keep it all a single target, then you can't add new features when you need them because you are being held back by the lowest framework you support.

We have plans to build a target for .NET Core and as a result, the nuget package *will* have more than one assembly in it.  Otherwise it's just not possible to have one assembly that fits .NET and .NET Core.

**Comment by phantomtypist** [comment]
I think this change works in .NET 2/3.x, but only by coincidence.  I'd have to do some testing, but I think because the .NET 2/3.x runtime doesn't have this new bug introduced in the .NET 4.x runtime, the cache expiration is just getting pushed out further... but not as obvious.

**Comment by phantomtypist** [comment]
Docs on System.Runtime.Caching: https://msdn.microsoft.com/en-us/library/system.runtime.caching(v=vs.110).aspx

See remarks at the bottom of the doc page.

**Comment by phantomtypist** [comment]
MS suggests that applications targeting .NET 4.x use System.Runtime.Caching instead of System.Web.Caching.  System.Web.Caching should be used in .NET 2/3.x applications.

https://docs.microsoft.com/en-us/dotnet/framework/performance/caching-in-net-framework-applications

**Comment by hokiecoder** [comment]
@phantomtypist No, I have not tested against .NET 2/3, in all honesty because I have no use cases that require those older framework versions. I had (rather naively) assumed that if it worked in version 4.X (not 4.7) then all was well. The proposed patch in pull request #55 needs to be checked against all versions of the framework that are supported by this project. If I can find some free time, I'll do that.

Regarding the dual targets, I think I misread what your purpose was. I thought you wanted two targets for the issue at hand: the caching expiration problem. But since that can be fixed easily by using UTC dates, it would be overkill. I now see that you want to do it for the `System.Web.Caching` vs `System.Runtime.Caching` idea. That makes sense to me. That being said, I think a new issue needs to be created to discuss and tackle since it is not directly related to this issue. If you want to create an issue for it and include all of the details you've mentioned here, mention me in the issue and I'll be happy to add my own thoughts!

**Comment by serac** [comment]
@phantomtypist I think you made a clear and compelling argument for multiple framework targets in the Nuget packages. +1 to that approach.

**Comment by phantomtypist** [comment]
Please see my comments in issue https://github.com/apereo/dotnet-cas-client/issues/56

**Comment by sbubaron** [comment]
I'm in agreement with the multi-platform approach.

**Comment by nicswan** [comment]
As a note, this issue manifested for me because .NET 4.7 was pushed as a critical update last week so was automatically added to production servers (yay!). So I imagine this will quickly become an issue for others as well. Thanks to @hokiecoder/@danransom for the fix. I was able to roll my own version and fix the issue for at least one of my apps.

**Comment by phantomtypist** [comment]
Heads up everyone!

We now have a pre-release NuGet package feed for alpha/beta/unstable versions of the DotNetCasClient package.  It can be found here: https://www.myget.org/gallery/dotnetcasclient-prerelease

Can you guys/gals please connect to the new feed and test the **1.1.0-beta0001** version of the package to see if this resolves the problem for you?  Report back if you have any problems or if the issue was resolved for you.

There were some other changes in this pre-release version, those can be found in the ReleaseNotes.md file found in the **release/1.1.0** branch right now.

Thanks!

@hokiecoder @danransom @webMan1 @serac @sbubaron @nicswan @scottt732 @mmoayyed

**Comment by jsaley** [comment]
Hi Jason, this works for us at campus residences. Thanks.

**Comment by nicswan** [comment]
@phantomtypist This is working for me on a 4.0.3 app on a server with .NET 4.7.

**Comment by MeliDreamingSand** [comment]
@phantomtypist This does not work for me. I couldn't get the logging to work either. I had to go back to the stable version to get the log to work again. 
According to the log from the stable version, the cache was not stored and the assertion expiration time is not correct.

**Comment by phantomtypist** [comment]
@MeliDreamingSand a few questions:

- What version of the "stable" NuGet package did you revert back to?
- Could you provide code on how you had the logging wired up when it wasn't working?
- What version of the .NET framework is your application targeting?
- What flavor of ASP.NET (and version) is your application using?

**Comment by MeliDreamingSand** [comment]
@phantomtypist thanks for the reply!

- I reverted back to the one that is currently on the master branch. 
- I copied-paste the logging directly from the release README. I was able to have the logging using that same code for the stable version.
- 4.5
- I'm using MVC 5 for the apps.

Please let me know if I need to provide more information. Thanks again!

**Comment by phantomtypist** [comment]
@MeliDreamingSand 

Still confused with one item.  When you say "I reverted back to the one that is currently on the master branch" did you mean:

1. You cloned the repo (master branch) and manually referenced the code/project in your project,
2. You referenced the nuget package version 1.0.2 from here (https://www.nuget.org/packages/DotNetCasClient/), or
3. You manually downloaded the release from here (https://github.com/apereo/dotnet-cas-client/releases/tag/v1.0.2) and manually referenced the DLL files in your project.

I only ask because there are no nuget packages stored in the repo nor master branch.

**Comment by MeliDreamingSand** [comment]
@phantomtypist 
Sorry for not making it clear. I manually downloaded the release from here (https://github.com/apereo/dotnet-cas-client/tree/master) and manually referenced the DLL files for debugging.
Currently all of our apps that's using the nuget package version 1.0.2 have the redirecting issue.

**Comment by sbubaron** [comment]
@MeliDreamingSand have you tried the pre-release version at https://www.myget.org/gallery/dotnetcasclient-prerelease ?

**Comment by MeliDreamingSand** [comment]
@sbubaron 
Yes I have tried that; it also didn't work.

**Comment by sbubaron** [comment]
Is this issue happening on your dev machines? Are you running Windows 10 with Creators Update?
Or are you running .net 4.7?

**Comment by MeliDreamingSand** [comment]
This issue is happening on both dev machine and production server. My dev machine is windows 10 with the .net 4.7 update while production server is windows 2008 with the .net 4.7 update as well.

**Comment by MeliDreamingSand** [comment]
@phantomtypist and @sbubaron 
Thank you all for helping! I am able to get the apps to work after deleting the .net 4.7 update, re-update it, and re-download the pre-release version.
Still can't get the logging to work, but at least the apps are working.
Thanks again!

**Comment by blakepell** [comment]
The individual critical 4.7 update caused this behavior for us also on a production server.  However, on a staging server that update was applied via ".NET Framework July 2017 Preview of Quality Rollup" with other patches, and that server did not exhibit the same behavior (presuming that it later applied something that fixed it).

Regardless, the 1.1.0-beta0001 branch fixed the issue.  Thank you to those that identified this and worked on it.

**Comment by terrybj** [comment]
Hi, anyone try this on Windows Server 2012 + .Net Framework 4.7? I still had such problem after getting NuGet package from https://www.myget.org/gallery/dotnetcasclient-prerelease.
Not sure if need more configuration? I just copied the components under net45 folder to website bin folder.

**Comment by phantomtypist** [comment]
@MeliDreamingSand 

I want to follow-up on the logging issue you are having.  Nothing was changed between 1.0.2 and the current beta for 1.1.0 with regards to logging.

Just to double check things... you have it configured like stated on the readme file in the repo (towards the bottom) and also took note of this note:

> The application pool in which the CAS-enabled .NET application runs must execute under a user with permission to create and write the trace log file.

If the application pool does have access to where you are writing the log file to, but your configuration is different, would you mind providing your logging configuration section so I can reproduce?  Alternatively, if you can make a sample application that fully reproduces the "no logging" and then share the repo so I can clone it and run it... that would be great.

Otherwise, I am not able to reproduce the "no logging" problem from my end on multiple machines :-/

**Comment by sbubaron** [comment]
@MeliDreamingSand 

Just to add on to @phantomtypist thoughts, my trace files aren't being written where I expected them to be, I used the default config specified so I was looking for them under C:\inetpub\logs\LogFiles

However, on IISExpress (my local dev box) they are instead being written to: C:\Users\USERNAME\Documents\IISExpress\TraceLogFiles\PROJECTNAME

Unsure what is causing this

**Comment by phantomtypist** [comment]
@MeliDreamingSand 

I believe I have fixed the logging issue.  Can you update your nuget package reference to the new 1.1.0-beta0004 release and let me know if that fixes logging for you?

https://www.myget.org/feed/dotnetcasclient-prerelease/package/nuget/DotNetCasClient

**Comment by sbubaron** [comment]
version 1.1.0-beta0004 worked for me, a dotnetcasclient.log is now being written to the file specified in web.config as expected. The files previously mentioned by me that are in the iisexpress\tracelogfiles\project are different (though contain some references to dotnetcas)

thanks @phantomtypist

**Comment by MeliDreamingSand** [comment]
@phantomtypist 
Version 1.1.0-beta0004 worked for me too.
Thank you!

**Comment by phantomtypist** [comment]
Great.  I think this rounds out all the issues people discussed during the beta testing so far.

I will be releasing the full 1.1.0 release shortly both on here as a binary release and also on NuGet.org (the production nuget feed for this package.)

**Comment by phantomtypist** [comment]
Version 1.1.0 released to nuget.org
